### PR TITLE
Support for user space docker

### PR DESCRIPTION
**Pull request by geetduggal**
@mr-c Thanks for taking a look!  Here is the code for optionally using a user space docker implementation that accepts relatively straightforward use of `pull` and `run` subcommands.  It effectively adds a `--user-space-docker-cmd` option and relaxes some constraints to use Docker in a very simplified way.  I tested it with our user-space solution (https://github.com/dnanexus/dx-toolkit/blob/master/src/python/scripts/dx-docker) but we would still need to run some more comprehensive test suite is run to ensure that the introduction of this doesn't break any existing functionality (in theory it shouldn't 👼 ).  I've attached some screenshots to illustrate its behavior:

e3af6e1a-4423-4bc8-827f-67dbd0d43558

4353596b-a8aa-4b85-b94b-ef2288cbf852

@chapmanb interest.

- [ ] add installation instructions for `dx-docker` and `udocker` to the README with example usage
- [ ] fix udocker's cmdline parsing, or document its limitations: https://github.com/indigo-dc/udocker/issues/87
- [x] fix dx-docker's cmdline parsing, or document its limitations: https://github.com/dnanexus/dx-toolkit/issues/252

### CODE REVIEW

**Comment by mr-c** [comment]
Jenkins, okay to test

File: cwltool/job.py [hunk 1]
```diff
@@ -360,32 +360,39 @@ def run(self, pull_image=True, rm_container=True,
...
-            if docker_is_req:
-                raise UnsupportedRequirement(
-                    "Docker is required to run this tool: %s" % e)
-            else:
-                raise WorkflowException(
-                    "Docker is not available for this tool, try --no-container"
-                    " to disable Docker: %s" % e)
+        user_space_docker_cmd = kwargs.get("user_space_docker_cmd")
+        if docker_req and user_space_docker_cmd:
+            img_id = str(docker_req["dockerPull"])
```
Comment (geetduggal) [inline cwltool/job.py]: Just realized this should also allow for `dockerImageId` for local or private images.

**Comment by geetduggal** [comment]
Added some minor changes -- will babysit the tests and fix anything if necessary

**Comment by geetduggal** [comment]
@mr-c Friendly ping on this, thanks in advance!

**Comment by mr-c** [comment]
@geetduggal Thank you for the reminder. I'm running the conformance tests using this and [`udocker`](https://github.com/indigo-dc/udocker) right now and making fixes with them.

Can you run the conformance tests using `dx-docker`?
1. Clone https://github.com/common-workflow-language/common-workflow-language/
2. `./run_tests.sh RUNNER=cwltool EXTRA="--user-space-docker-cmd=dx-docker"
3. If not all tests pass, modify `dx-docker` as needed

**Comment by mr-c** [comment]
Currently both `udocker` and `dx-docker` fail conformance tests 66 & 67

File: cwltool/job.py [hunk 2]
```diff
@@ -394,25 +407,30 @@ def run(self, pull_image=True, rm_container=True,
...
-        if not onWindows():  # MS Windows does not have getuid() or geteuid() functions
-            euid, egid = euid or os.geteuid(), egid or os.getgid()
+            euid, egid = docker_vm_id()
+            if not onWindows():  # MS Windows does not have getuid() or geteuid() functions
+                euid, egid = euid or os.geteuid(), egid or os.getgid()
 
-        if kwargs.get("no_match_user", None) is False and (euid, egid) != (None, None):
-            runtime.append(u"--user=%d:%d" % (euid, egid))
+            if kwargs.get("no_match_user", None) is False and (euid, egid) != (None, None):
+                runtime.append(u"--user=%d:%d" % (euid, egid))
```
Comment (mr-c) [inline cwltool/job.py]: `udocker` does support `--user`; can `dx-docker` just ignore this?

Comment (geetduggal) [reply cwltool/job.py]: Yes, good point. I'll add a feature to it to ignore it for now and add support down the line.  Here's the PR I have for ignoring: https://github.com/dnanexus/dx-toolkit/pull/254

**Comment by geetduggal** [comment]
@mr-c Thanks for looking into this!   I submitted a PR for our docker implementation to ignore the user command and am running the tests here.  (@chapmanb interest.)

**Comment by geetduggal** [comment]
Sorry I missed the Issue posted to the toolkit -- thanks!  I am testing it out as I write this, and the conformance tests have succeeded now:

image

**Comment by geetduggal** [comment]
Also added an inital draft for documentation

**Comment by mr-c** [comment]
@geetduggal Can you merge https://github.com/geetduggal/cwltool/pull/1 ?

**Comment by geetduggal** [comment]
Done

**Comment by geetduggal** [comment]
@mr-c I *think* we are good to go on the `dx-docker` side here (the files are all merged into the master branch of our toolkit).  Does it seem reasonable on your end?   Thanks in advance!

**Comment by mr-c** [comment]
@cwl-bot A little late to the game!

**Comment by geetduggal** [comment]
😂

**Pull request merged**